### PR TITLE
When playing back an item that is an addon pass a new script param to…

### PR DESCRIPTION
**Pull request by faush01**
… addon script that lets the addon know if the user selected resume of not in the built in resume dialog question.

If a user selected resume then the string resume will be passed to the addon script.

## Description
When the user plays an item that is an addon item and there is a progress set they will be given a resume dialog. All this change does is pass the result of the resume question on to the addon via a new script param string "resume".
This allows addons to use the built in resume question response instead of having to handle this in the addon, the addon will have the answer to what the user wanted to do.

## Motivation and Context

I have written a number of addons over the years and this allows you to build addons that are more in line with the global look and feel of the app.
Having this data passed to the addon script gives the user a smoother experience.

## How Has This Been Tested?

Tested with addons and playback process to test regressions.

Tested on Windows, Run all the built in tests.

This change is low risk, none of the internal logic is being changed, if a user asks for resume a new script param is added when the addon is called, that is all.
## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

**Comment by notspiff** [comment]
uhm, you can read the startoffset in the add-on can't you ? that was at least the original intent.

**Comment by faush01** [comment]
That makes sense, how is it surfaced in the listitem? As a property?

**Comment by notspiff** [comment]
it is available through getProperty('startoffset') indeed. but yes, i now recall how it works;

there are two ways to do this: one is using setResolvedUrl type operation. in that case, if the *original* url (i.e. not the resolved media url!) is a constant, kodi will save info based on the original url and handle the resume for that item as normal.

alternatively you can set the startoffset on the resolved item returned in the add-on. in that case kodi will use whatever value you put in there (using setProperty('startoffset', secs-as-float)). the second here was intended to be the hook used if you want to handle this on the add-on side.

**Comment by faush01** [comment]
setResolvedUrl  might work but the addon I am writing does not use this approach.

The addon uses the following approach:
 - Get a list of media items from the Emby Media server https://emby.media
 - Build a list of ListItems and return that to Kodi to be displayed
 - In the ListItems the progress is set so Kodi and the Kodi skins show the correct progress icons and indicators
 - The URL of the ListItems are set to a plugin:// play action in the addon so when a user selects an item the addon is called with the ID of the item to play
 - Because the progress is set in the ListItem when a user selects an in progress item they will be promoted for resume or start from beginning by the default built in Kodi resume question
 - The play action of the addon is called and this is responsible for getting the actual URL and using the kodi player to play the file
 - If the item is part played (details from the Emby Server) then seek is called on the playback item to take the player to the users progress point

I the above scenario the only way I can get access to the answer to the built in resume or from start question is using the code change I have submitted.

This should be all low risk as the only logic change is passing along to the addon script the extra param of resume:true|false

**Comment by faush01** [comment]
It's coming up to the weekend here so if anyone has any feedback or changes let me know and I will spend some time over the next few days sorting them out.

Otherwise if no feedback are we happy to merge this?

**Comment by faush01** [comment]
Any more feedback?
Can we merge this?

**Comment by faush01** [comment]
Hi guys, not sure what you want to do with this one now, its been a while, just wondering if you want to merge or just close it.

File: xbmc/filesystem/PluginDirectory.cpp [hunk 1]
```diff
@@ -151,6 +151,11 @@ bool CPluginDirectory::StartScript(const std::string& strPath, bool retrievingDi
   argv.push_back(strHandle);
   argv.push_back(options);
 
+  std::string strResume = "resume:false";
+  if (resume)
+    strResume = "resume:true";
+  argv.push_back(strResume);
+
   // run the script
   CLog::Log(LOGDEBUG, "%s - calling plugin %s('%s','%s','%s')", __FUNCTION__, m_addon->Name().c_str(), argv[0].c_str(), argv[1].c_str(), argv[2].c_str());
```
Comment (Razzeee) [inline xbmc/filesystem/PluginDirectory.cpp]: You might want to also log it here

File: xbmc/filesystem/PluginDirectory.cpp [hunk 2]
```diff
@@ -464,6 +469,11 @@ bool CPluginDirectory::RunScriptWithParams(const std::string& strPath)
   argv.push_back(strHandle);
   argv.push_back(options);
 
+  std::string strResume = "resume:false";
+  if (resume)
+    strResume = "resume:true";
+  argv.push_back(strResume);
+
   // run the script
   CLog::Log(LOGDEBUG, "%s - calling plugin %s('%s','%s','%s')", __FUNCTION__, addon->Name().c_str(), argv[0].c_str(), argv[1].c_str(), argv[2].c_str());
```
Comment (Razzeee) [inline xbmc/filesystem/PluginDirectory.cpp]: Missing log again

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 3]
```diff
@@ -763,7 +763,7 @@ std::string CGUIWindowVideoBase::GetResumeString(const CFileItem &item)
 
 bool CGUIWindowVideoBase::ShowResumeMenu(CFileItem &item)
 {
-  if (!item.m_bIsFolder && !item.IsPVR())
+  if (!item.m_bIsFolder && !item.IsPVR() || item.IsPlugin())
```
Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: This seems weird and/or dangerous. A plugin should already confirm to the first two rules. It's not a folder and it's not PVR.

Comment (faush01) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: yes, just confirmed, item.m_bIsFolder is True when an addon is called from the Info screen.

Comment (Rechi) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: m_bIsFolder depends on the value specified for isFolder at  xbmcplugin.addDirectoryItem (https://codedocs.xyz/xbmc/xbmc/group__python__xbmcplugin.html#ga4c357cf7d3dbd4e2823755ecfd7dfa01)

Comment (faush01) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: I am using:

dirItems is an array of (u, list_item, folder)
xbmcplugin.addDirectoryItems(pluginhandle, dirItems)

and folder is being set correctly, it is false for the items I am adding.

This is occurring when I hit the play button on the media info screen. It does not happen when I select play for the item in the dir list, but if I hit Info to show info screen and then hit the play this is when I see item.m_bIsFolder being True

Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: Then that's a bug. Nothing you should try to tackle in this part of the code :)

Comment (faush01) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: Ok, I will remove this change.
So do you want me to try to find the root cause for this?

Comment (Razzeee) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: If you can put time into that, that would be appreciated.

Comment (faush01) [inline xbmc/video/windows/GUIWindowVideoBase.cpp]: The root cause of this was the new movie item not being created from the original CFileItemPtr but it used the video info tag to populate it. This used some weird logic to work out if it was a folder or not. If there was no FileAndPath set in the video info tag then it was assumed it was a dir and set the path to / and marked it as a Folder.

I made a small change to populate the movie item from the base CFileItemPtr, tjis uses the original items IsFolder value.

This is now working.

**Comment by Razzeee** [comment]
@notspiff I would like to hear your thought on this, with the added information :)

ios and linux failed, let's try again

jenkins build this please

**Comment by Rechi** [comment]
please rebase first, before trying to build it again

**Comment by Razzeee** [comment]
whoops, got ahead of myself

**Comment by akva2** [comment]
sure but i'm busy busy, monday at the earliest.

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp [hunk 4]
```diff
@@ -597,7 +597,7 @@ void CGUIDialogVideoInfo::Play(bool resume)
     return;
   }
 
-  CFileItem movie(*m_movieItem->GetVideoInfoTag());
+  CFileItem movie(*m_movieItem);
```
Comment (notspiff) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: this looks potentially icky. it will change from an item pointing to the actual underlying file to an item pointing to a library item. not sure if it has consequences but why is this change needed?

**Comment by notspiff** [comment]
it seems reasonable. only thing i question is whether the bool could be parsed more elegantly somehow. a lot of signature impact. it for sure could be given a default of false in any case.

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp (hunk 4 above)
Comment (faush01) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: There are overloaded constructors, the one that takes a VideoInfoTag does not populate all the member level isFolder type data it uses some internal logic to set them from data in the VideoInfoTag which is not consistent with what the original item has.
Using the base Item it allows the new movie item to be populated using the full item info.
I feel it is actually more correct to use the constructor that passes in the full item if it is available, in this instance the full item is available and should be used.

Comment (notspiff) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: i know for a fact that 5 years ago i tracked a bug down to the fact that we did what you suggest, and changed it to what it is now. i cannot guarantee whether the problem still exists but this potentially causes regressions and i this needs to be investigated! i do not recall the details..

Comment (notspiff) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: actually it's more like 10 years ago, sometimes prior to 2009 so it's all smashed into the svn to git conversion ;) but i looked into it. the problem is that we hop into the ShowResumeMenu, which calls into the Database to get resume points. these are stored by path, and not by the virtual db path, so i think that was the issue. but then again maybe the surrounding code has changed so the m_movieItem points to proper paths in the first place now?

Comment (faush01) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: I had a look back through the git log for this file but I could only go back to 25/1/2011 and could not see any changes to this function.
I am not sure what or how to regression test this with no idea where to start with what to look for.

Comment (faush01) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: ok so its a DB path resume point issue, I can test this.

Comment (faush01) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: the other option is just to use the first approach and check if the item is a plugin when displaying the resume dialog.

Comment (notspiff) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: sounds safer. but current code is non-intuitive so if it can be changed that is preferable.

Comment (faush01) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: So the original idea was to add
if (!item.m_bIsFolder && !item.IsPVR() **|| item.IsPlugin()**)
to
xbmc/video/windows/GUIWindowVideoBase.cpp 
See above in the hidden change request, it was requested that I dont do that there and find the root cause. 
I found the root cause as the IsFolder not being set so then is when I realised the IsFolder was not being set when using the VideoInfoTag constructor.
If you think it is better to go back to the original idea of adding || item.IsPlugin() then I can do that.

**Comment by faush01** [comment]
Ok I changed it back to use the original approach of using item.IsPlugin() in the if statement.
I am really not sure where to go from here, if we have people unhappy with both approaches perhaps someone can suggest a new approach.
Or we can just go with this approach as it is quite descriptive, if it is a plugin then show the resume, cant get more descriptive than that :-)

**Comment by notspiff** [comment]
rest assured we will find a solution. nobody is objecting your change, it's just takes some stamina to work in huge code bases :)

File: xbmc/video/dialogs/GUIDialogVideoInfo.cpp (hunk 4 above)
Comment (Razzeee) [inline xbmc/video/dialogs/GUIDialogVideoInfo.cpp]: @faush01 did you find some time to go back and look at "ok so its a DB path resume point issue, I can test this."?
might be the best way to proceed, otherwise we'll have to live with the "IsPlugin-Hack" But it will need to be documented/commented.

**Comment by faush01** [comment]
So ultimately the root cause is that the VideoInfoTag did not have a m_strFileNameAndPath set.
To allow an addon to set this I added a info tag to the video type on the ListItem.cpp to allow an addon to set FileNameAndPath, this fixes the root cause, is backwards compatible and should have the lowest code impact of all.
What do people think?

**Comment by Razzeee** [comment]
should work fine, as long as addons make sure to always set it consistently, but that's out of our hands obviously :)
sounds fine to me (without seeing any code)

**Comment by faush01** [comment]
I have amended the changes in this pull request to include the video info tag FileNameAndPath already, you should be able to see the code in the change set in this pull request.

**Comment by faush01** [comment]
I have re-based on master to fix a conflict with the changes on master since this pull request has been open for so long.
It would we good if we could nail this down soon.

**Comment by faush01** [comment]
Hi Guys, its getting close to 2 months this has been open.
I think we have covered all the requests for change and found the lowest risk and most compatible solution.
I was just wondering if this can be merged now.
Thanks.

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by faush01** [comment]
Not sure why the failed build, the failed build logs look like they been deleted now.

**Comment by Razzeee** [comment]
jenkins build this please

**Comment by faush01** [comment]
Anyone know what this error means:

```
Checking out Revision 45d0cd49da1028f714e6b2bde920a3c0695b8f0c (refs/remotes/origin/pr/12181/merge)
org.eclipse.jgit.errors.MissingObjectException: Missing unknown 45d0cd49da1028f714e6b2bde920a3c0695b8f0c
```

**Comment by MartijnKaijser** [comment]
Jenkins build this please

**Comment by MartijnKaijser** [comment]
@faush01 could you please do a rebase on latest master and force push?

**Comment by faush01** [comment]
@MartijnKaijser I have rebased on latest master

**Comment by faush01** [comment]
Jenkins build this please

**Comment by Razzeee** [comment]
jenkins build this please
